econml/dml.py

```diff
@@ -95,15 +126,26 @@ def _fit_impl(self, Y, T, X=None, W=None):
             # TODO: If T is a vector rather than a 2-D array, then the model's fit must accept a vector...
             #       Do we want to reshape to an nx1, or just trust the user's choice of input?
             #       (Likewise for Y below)
-            self._models_t[idx].fit(X_train, W_train, T_train)
+            if sample_weight is not None:
+                self._models_t[idx].fit(X_train, W_train, T_train, sample_weight=sample_weight[train_idxs])
+            else:
+                self._models_t[idx].fit(X_train, W_train, T_train)
             if self._discrete_treatment:
                 T_res[test_idxs] = T_test - self._models_t[idx].predict(X_test, W_test)[:, 1:]
             else:
                 T_res[test_idxs] = T_test - self._models_t[idx].predict(X_test, W_test)
```

Comment:
We need to fix the shape mismatch that is created by the sklearn models here, in this merge. I tried the following code and get the following error:
```python
# DGP constants
np.random.seed(123)
n=1000
n_w=30
support_size=5
n_x=1
# Outcome support
support_Y=np.random.choice(range(n_w),size=support_size,replace=False)
coefs_Y=np.random.uniform(0,1,size=support_size)
epsilon_sample=lambda n: np.random.uniform(-1,1,size=n)
# Treatment support
support_T=support_Y
coefs_T=np.random.uniform(0,1,size=support_size)
eta_sample=lambda n: np.random.uniform(-1,1,size=n)

# Generate controls, covariates, treatments and outcomes
W=np.random.normal(0,1,size=(n,n_w))
X=np.random.uniform(0,1,size=(n,n_x))
# Heterogeneous treatment effects
TE1=np.array([exp_te(x_i) for x_i in X])
TE2=np.array([x_i**2 for x_i in X]).flatten()
T=np.dot(W[:,support_T],coefs_T)+eta_sample(n)
Y=TE1*T+TE2*T**2+np.dot(W[:,support_Y],coefs_Y)+epsilon_sample(n)
# Generate test data
X_test = np.array(list(product(np.arange(0, 1, 0.01), repeat=n_x)))
est = LinearDMLCateEstimator(model_y=RandomForestRegressor(),
                             model_t=RandomForestRegressor(),
                             featurizer=PolynomialFeatures(degree=2))
T=T.reshape(-1,1)
est.fit(Y, T, X, W)
```

Error message:
```
c:\users\vasy\documents\econml\econml\dml.py:294: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self._model.fit(self._combine(X, W), Target)


---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-118-c754508b77f5> in <module>
      1 T=T.reshape(-1,1)
----> 2 est.fit(Y, T, X, W)

c:\users\vasy\documents\econml\econml\dml.py in fit(self, Y, T, X, W, sample_weight, inference)
     98         assert shape(Y)[0] == shape(T)[0] == shape(X)[0] == shape(W)[0]
     99 
--> 100         Y_res, T_res = self.fit_nuisances(Y, T, X, W, sample_weight=sample_weight)
    101 
    102         self.fit_final(X, Y_res, T_res, sample_weight=sample_weight)

c:\users\vasy\documents\econml\econml\dml.py in fit_nuisances(self, Y, T, X, W, sample_weight)
    134                 T_res[test_idxs] = T_test - self._models_t[idx].predict(X_test, W_test)[:, 1:]
    135             else:
--> 136                 T_res[test_idxs] = T_test - self._models_t[idx].predict(X_test, W_test)
    137             if sample_weight is not None:
    138                 self._models_y[idx].fit(X_train, W_train, Y_train, sample_weight=sample_weight[train_idxs])

ValueError: shape mismatch: value array of shape (500,500) could not be broadcast to indexing result of shape (500,1)
```

Reply:
When the treatment is multi-dimensional then things are fine with effect estimation. However things break with effect_interval estimation using statsmodels. For instance, if we change the last lines to:
```python
T=T.reshape(-1,1)
est.fit(Y, np.concatenate((T, T**2), axis=1), X, W, inference='statsmodels')
te_pred=est.effect(X_test)
te_pred_int = est.effect_interval(X_test, alpha=.01)
```

Then we get the following error message:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-126-48e3f8843d91> in <module>
      1 te_pred=est.effect(X_test)
----> 2 te_pred_int = est.effect_interval(X_test, alpha=.01)

c:\users\vasy\documents\econml\econml\dml.py in effect_interval(self, X, T0, T1, alpha)
    197         return super().effect_interval(X,
    198                                        T0=self._expand_treatment(T0, X), T1=self._expand_treatment(T1, X),
--> 199                                        alpha=alpha)
    200 
    201     def score(self, Y, T, X=None, W=None):

c:\users\vasy\documents\econml\econml\cate_estimator.py in call(self, *args, **kwargs)
    124             name = m.__name__
    125             if self._inference is not None:
--> 126                 return getattr(self._inference, name)(*args, **kwargs)
    127             else:
    128                 AttributeError("Can't call '%s' because 'inference' is None" % name)

c:\users\vasy\documents\econml\econml\inference.py in effect_interval(self, X, T0, T1, alpha)
    106     def effect_interval(self, X, *, T0=0, T1=1, alpha=0.1):
    107         return self.props.effect_op.apply(self.props.wrapper.predict_interval, (X, T1 - T0),
--> 108                                           interval=True, make_const_marginal_effect=False, alpha=alpha)
    109 
    110     def const_marginal_effect_interval(self, X, *, alpha=0.1):

c:\users\vasy\documents\econml\econml\dml.py in apply(self, predict, input, interval, make_const_marginal_effect, *args, **kwargs)
     35             # to each input row
     36             input = broadcast_unit_treatments(input, self._d_t[0] if self._d_t else 1)
---> 37         preds = predict(self._pre(input), *args, **kwargs)
     38         if make_const_marginal_effect:
     39             # for constant marginal effects, we need to reshape the results to ensure

c:\users\vasy\documents\econml\econml\utilities.py in predict_interval(self, X, alpha)
    840             X = add_constant(X, has_constant='add')
    841         # NOTE: we use `obs = False` to get a confidence, rather than prediction, interval
--> 842         preds = self.results.get_prediction(X).conf_int(alpha=alpha, obs=False)
    843         # statsmodels uses the last dimension instead of the first to store the confidence intervals,
    844         # so we need to transpose the result

C:\ProgramData\Anaconda3\lib\site-packages\statsmodels\regression\linear_model.py in get_prediction(self, exog, transform, weights, row_labels, **kwds)
   2308         return pred.get_prediction(
   2309             self, exog=exog, transform=transform, weights=weights,
-> 2310             row_labels=row_labels, **kwds)
   2311 
   2312     get_prediction.__doc__ = pred.get_prediction.__doc__

C:\ProgramData\Anaconda3\lib\site-packages\statsmodels\regression\_prediction.py in get_prediction(self, exog, transform, weights, row_labels, pred_kwds)
    159     if pred_kwds is None:
    160         pred_kwds = {}
--> 161     predicted_mean = self.model.predict(self.params, exog, **pred_kwds)
    162 
    163     covb = self.cov_params()

C:\ProgramData\Anaconda3\lib\site-packages\statsmodels\regression\linear_model.py in predict(self, params, exog)
    343             exog = self.exog
    344 
--> 345         return np.dot(exog, params)
    346 
    347     def get_distribution(self, params, scale, exog=None, dist_class=None):

ValueError: shapes (100,3) and (6,) not aligned: 3 (dim 1) != 6 (dim 0)
```

Reply:
This works if you specify `T0` and `T1` explicitly, but it should also work if they're implicit since the non-interval code works there.  I'll see if there's a good way to plumb things through.

Reply:
There seems to be a bug in the non-interval code too. Why is est.effect(X_test) returning an object of size (100, 2). effect(X_test) should be a scalar for each sample. So it should have a size (100,). const_marginal_effect and marginal_effect should have size (100, 2), but not effect.

Reply:
This seems to be related with how we treat default T0, T1. I think we definitely need to replicate the 0 and the 1 to the same dimension as the training shape. I don't see why we didn't do that. If we do that then effect in the above code will return correctly a vector and even effect interval will work.

Reply:
@moprescu I checked the ortho forest notebook and in the example with the multiple treatments, effect(X) also returns a (n, n_treatments) result, as opposed to (n, ) is that a bug in the ortho forest effect code?

Reply:
@vasilismsr Historically, we didn't do it because the expansion is defined in the `LinearCateEstimator` class, which doesn't track the dimensionality of `T` that was passed to `fit` (because all of the fitting logic goes in subclasses), so there was no way to restore it in `effect`.

In the time since that logic was originally put into place, the base `BaseCateEstimator` class has gotten a fit implementation (to take care of inference), so we could now put in a line that stores the trailing dimensions of `T` (if any).

On the other hand, with multiple treatments I'm not sure how sensible it is to propagate the defaults in the first place (why would you care only about switching _all_ treatments from 0 to 1 _simultaneously_?).  Arguably it would be better to force the user to specify the treatment arguments explicitly in the multiple treatment case.

Reply:
(But if want the user to specify the multiple treatments, the behavior should be changed to throw a helpful error if they aren't passed in, not compute meaningless garbage as the non-interval case does now...)

Reply:
Yeap. To the least we need some error message as now it was very prone to a user error (as what happened to me, where I did not specify a T0, T1 and I should have known to specify because I have multi-dim treatment).

One proposal: 
1) keep track of dim(T) at the fit of the BaseCate estimator and set the defaults to going from all zeros to all ones.
2) if treatment is multi-dim then at effect(X) time, if the defaults T0, T1 are used then throw a warning that says sth like: "You have multiple treatments and called effects with default T0, T1, which calculates effect from going from all zero treatments to all one treatments. Consider setting your baseline and target treatment vectors. For the marginal effect vector consider using the marginal effect or const_marginal_effect methods." Or sth more concise.

Reply:
One problem might be: what if some developer creates an estimator, but does not call the super().fit() at the end of fit? This would be ok for all purposes except for the case of multi-dim treatment. We need to specify somewhere in the base classes that every inherited class must call super().fit() in the end.

Reply:
Why don't we implement sth like the following: we can create a fit_decorator (maybe even an effect_decorator, a const_marginal_effect decorator etc. for every main method). Then the developer-user has to always decorate their methods with this decorator. Then we can implement whatever pre and post processing logic we want in our basecase estimator python file, where we define the decorator. This way the developer of subclasses does not need to know what pre and post processing needs to happen and moreover we can change the pre and post processing logic at the highest level and in one place. For instance, here is an example of a post-processing I was imagining. The example below will always call the super().fit() at the end of the inherited class:
```python
def fit_decorator(fit):
    def wrapper(*args, **kwargs):
        obj = fit(*args, **kwargs)
        new_args=args[1:]
        super(type(args[0]), args[0]).fit(*new_args, **kwargs)
        return obj
    return wrapper

class Super():
    def __init__(self):
        return
    def fit(self):
        print("Super")
        return self

class Sub(Super):
    def __init__(self):
        return
    @fit_decorator
    def fit(self):
        print("Sub")
        return self

Sub().fit()
```

Reply:
I tried it and something like the following decorator works:
```python

def fit_decorator(fit):
    def wrapper(*args, **kwargs):
        fit(*args, **kwargs)
        return BaseCateEstimator.fit(*args, **kwargs)
    return wrapper
```
Then we can remove the final call to super in each of the subclasses and only have to decorate the fit method. We could then also add to the above decorator any preprocessing the we want, which for instance could keep track of the dimension of T.

Reply:
Sth like the following:
```python
def fit_decorator(fit):
    def wrapper(*args, **kwargs):
        BaseCateEstimator._check_fit_input(*args, **kwargs)
        fit(*args, **kwargs)
        return BaseCateEstimator.fit(*args, **kwargs)
    return wrapper


class BaseCateEstimator(metaclass=abc.ABCMeta):
    """
    Base class for all CATE estimators in this package.
    """

    def _get_inference_options(self):
        """
        Produce a dictionary mapping string names to `Inference` types.

        This is used by the `fit` method when a string is passed rather than an `Inference` type.
        """
        return {'bootstrap': BootstrapInference}

    def _check_fit_input(self, Y, T, *args, **kwargs):
        self._dT = T.shape
```

Reply:
@vasilismsr I'm not sure that "you need to apply this decorator" is any better than "you need to call super" (see also the answers to https://stackoverflow.com/questions/34224896/force-child-class-to-call-parent-method-when-overriding-it; the top answer suggests just documenting the requirement while the second answer suggests something more like the `_fit_impl` approach used by a previous iteration of our code).

Reply:
In any case, the exact approach we take internally is something that we can change later without affecting user-facing behavior.

For now, I'll make a targeted change so that we can correctly handle 2D treatments along the lines of your warning suggestion above, but I'm going to postpone more aggressive refactoring for later work.  I don't want that to hold up the current set of changes which are mostly about enabling StatsModels inference and supporting sample weights.

Reply:
The reason why I like the decorator is two-fold: 
1) first we implement any pre and post processing logic in one place and when we make changes to any pre and post processing code, we only change that one place. For instance, now if we need to add some preprocessing, then we need to go over all fit methods and add some super call at the beginning of fit. If we had a decorator we wouldn't need to. 
2) A decorator does seem to me less invasive than a line of code in the method itself. 

Your suggestion for this PR seems good.

Reply:
For instance, with the decorator, it might be possible to push the "discrete treatment" preprocessing to happen inside the decorator and at one place.

Reply:
I agree that the ability to do pre- and post-processing is valuable, but I think that the Template Method suggestion [here](https://stackoverflow.com/a/34225912/82959) is probably a more idiomatic approach than a decorator.  But we can discuss further when we do the next round of refactoring.

Reply:
Hm. I think one thing that I don't like about this is documentation for the user: I find docstrings to be the most valuable, and if the child class only implements the private method, then there wont be a method docstring that is child class specific and visible to the user, while they are typing in some editor environment.